client/rpc/src/eth_pubsub.rs

```diff
@@ -251,164 +243,151 @@ where
+							{
+								Ok(txs) => {
+									if txs.len() == 1 {
+										Some(txs[0].clone())
+									} else {
+										None
 									}
-									_ => None,
-								};
-								futures::future::ready(res)
-							} else {
-								futures::future::ready(None)
-							}
-						})
-						.map(|transaction| {
-							return Ok::<Result<PubSubResult, jsonrpc_core::types::error::Error>, ()>(
-								Ok(PubSubResult::TransactionHash(transaction.hash())),
-							);
-						});
-					stream
-						.forward(
-							sink.sink_map_err(|e| warn!("Error sending notifications: {:?}", e)),
-						)
-						.map(|_| ())
-				});
+								}
+								_ => None,
+							};
+							futures::future::ready(res)
+						} else {
+							futures::future::ready(None)
+						}
+					})
+					.map(|transaction| PubSubResult::TransactionHash(transaction.hash()));
+
+				subscribe_helper(&self.subscriptions, sink, stream)
 			}
 			Kind::Syncing => {
-				self.subscriptions.add(subscriber, |sink| {
-					let mut previous_syncing = network.is_major_syncing();
-					let stream = client
-						.import_notification_stream()
-						.filter_map(move |notification| {
-							let syncing = network.is_major_syncing();
-							if notification.is_new_best && previous_syncing != syncing {
-								previous_syncing = syncing;
-								futures::future::ready(Some(syncing))
-							} else {
-								futures::future::ready(None)
-							}
-						})
-						.map(|syncing| {
-							return Ok::<Result<PubSubResult, jsonrpc_core::types::error::Error>, ()>(
-								Ok(PubSubResult::SyncState(PubSubSyncStatus {
-									syncing: syncing,
-								})),
-							);
-						});
-					stream
-						.forward(
-							sink.sink_map_err(|e| warn!("Error sending notifications: {:?}", e)),
-						)
-						.map(|_| ())
-				});
+				let mut previous_syncing = network.is_major_syncing();
+				let stream = client
+					.import_notification_stream()
+					.filter_map(move |notification| {
+						let syncing = network.is_major_syncing();
+						if notification.is_new_best && previous_syncing != syncing {
+							previous_syncing = syncing;
+							futures::future::ready(Some(syncing))
+						} else {
+							futures::future::ready(None)
+						}
+					})
+					.map(|syncing| PubSubResult::SyncState(PubSubSyncStatus { syncing: syncing }));
+
+				subscribe_helper(&self.subscriptions, sink, stream)
 			}
 		}
 	}
+}
 
-	fn unsubscribe(
-		&self,
-		_metadata: Option<Self::Metadata>,
-		subscription_id: SubscriptionId,
-	) -> JsonRpcResult<bool> {
-		Ok(self.subscriptions.cancel(subscription_id))
-	}
+fn subscribe_helper<S, Item>(
+	executor: &SubscriptionTaskExecutor,
+	mut sink: SubscriptionSink,
+	stream: S,
+) -> RpcResult<()>
+where
+	S: futures::stream::Stream<Item = Item> + Send + 'static,
+	Item: Send + sp_runtime::Serialize + 'static,
+{
+	let fut = async move {
+		stream
+			.take_while(|item| {
+				std::future::ready(sink.send(&item).map_or_else(
```

Comment:
[use the stream API instead](https://github.com/paritytech/jsonrpsee/blob/master/core/src/server/rpc_module.rs#L756-#L775)

this might leak if the never terminates see https://github.com/paritytech/jsonrpsee/issues/627